internal/info/info.go

```diff
@@ -54,43 +69,84 @@ type StackTrace struct {
 }
 
 var (
-	Version      = "v0.0.1"
-	GitCommit    = "master"
+	// injected from build script
+	Version           = "v0.0.1"
+	GitCommit         = "master"
+	BuildTime         = ""
+	GoVersion         string
+	GoOS              string
+	GoArch            string
+	CGOEnabled        string
+	NGTVersion        string
+	BuildCPUInfoFlags string
+
 	Organization = "vdaas"
 	Repository   = "vald"
-	BuildTime    = ""
 
-	GoVersion  string
-	GoOS       string
-	GoArch     string
-	CGOEnabled string
+	reps = strings.NewReplacer("_", " ", ",omitempty", "")
 
-	NGTVersion string
+	once         sync.Once
+	infoProvider InformationProvider
+)
 
-	BuildCPUInfoFlags string
+// Init initializes Detail object only once.
+func Init(name string) {
+	once.Do(func() {
+		infoProvider, _ = New(WithServerName(name))
+	})
+}
```

Comment:
Hmmmm.. 🤔 
I agree with your opinion.
but if you fix it, the interface will change, so it may be necessary to discuss it at first.

Reply:
yes I think so. what do you think?
@vankichi @hlts2

Reply:
I think it would be better retuning the error.
when you apply this implementation, how this change will affect another current implementations?

Reply:
@vankichi @hlts2 
changes of this function will cause the refactoring on 1 another file.
https://github.com/vdaas/vald/search?q=%22info.init%22

The image of the changes is like this, and this changes will cause the application will not be started if the empty string argument passed into this function. Or we can only log the error without returning the error here?
```go
	if err := info.Init(r.name); err != nil {
               // maybe log the error only? 
               return err
         } 
```

Reply:
What is this package needed for ?
I'd like to know how errors of `info.Init()` affects in Vald.

Reply:
@vankichi Since it will cause `runner` package changes, it will affect all of the Vald components.

Reply:
well, thanks for the reply.
`return error` is important to realize why the container is crashed.
please keep taking care of avoiding to include any bugs due to your refactor.

Reply:
@vankichi Sorry for the question again.
Here is one of the example of the caller implementation of the [runner](https://github.com/vdaas/vald/blob/133f0ab61013fffedd7e8795dc29d1e57a15d10b/internal/runner/runner.go#L65).
https://github.com/vdaas/vald/blob/133f0ab61013fffedd7e8795dc29d1e57a15d10b/cmd/meta/cassandra/main.go

If we return error and do not initialize the `info` instance from our implementation, the following panics/problems will be occurred in [line 55 of main.go](https://github.com/vdaas/vald/blob/133f0ab61013fffedd7e8795dc29d1e57a15d10b/cmd/meta/cassandra/main.go#L55).
1. info instance is nil
2. logger is not initialized

What should I do?

Reply:
and this refactoring cause this compilation error. I will think how to fix it.
https://github.com/vdaas/vald/pull/862/checks?check_run_id=1593219091

Source code: https://github.com/vdaas/vald/blob/master/internal/net/grpc/status/status.go#L71

Reply:
@kevindiu could you show us the options for resolve the problem which were discussed in SET team.
and please try to get some feedbacks from co-authors :)

Reply:
[Background]
When refactoring the info package, the `New()` function is implemented and used to initialize the singleton instance inside Init().
Because the `New()` returns error now, so it is better to handle/return the error inside `Init()` . The [current implementation of Init()](https://github.com/vdaas/vald/blob/b54d583b35b699a2d624d453b7d30ebf2f169113/internal/info/info.go#L94) returns error.

[Problem]
After the refactoring, I also refactor the caller side `runner.Do()` to handle the error, so I implemented [like this](https://github.com/vdaas/vald/blob/b54d583b35b699a2d624d453b7d30ebf2f169113/internal/runner/runner.go#L65), to return the error inside runner.Do().
But this implementation cause a problem that, the caller side of `runner.Do()` (for example `cmd/main.go`) will cause panic (nil pointer error) if the `info.Init()` failed and the singleton didn't initialized.
Code: https://github.com/vdaas/vald/blob/133f0ab61013fffedd7e8795dc29d1e57a15d10b/cmd/meta/cassandra/main.go#L55

[Solutions]
After some discussions in SET team, we found the there are 3 potential solutions for this problem.
1. `Init()`  do not return error and ignore the error inside the implementation
1. log the error inside runner.Do() and break the program
1. Exit the program in the Init when object creation fails

What do you think? @rinx @kpango 
cc: @hlts2 @vankichi

Reply:
I think the `Init()` functions should not return error, same as the current implementation of master branch.
If the `New() error` functions returns some error inside the `Init()` function, it means the arguments passed in `Init()` function is invalid = it means there're bugs in our codes. So it is better to use `log.Fatal` like the following:
```go
func Init(name string) error {
        ip, err = New(WithServerName(name)) // the passed arguments of the New function is just controlled by our code here!
        if err != nil {
            log.Fatal("info.Init failed")
        }
```
The behavior of the `Init()` function cannot be controlled by the config.yaml, I think it is okay to panic here.

Reply:
@rinx I am not sure if your suggestion works, because the logger `log.Init()` is executed after the `info` is initialized.
https://github.com/vdaas/vald/blob/b54d583b35b699a2d624d453b7d30ebf2f169113/internal/runner/runner.go#L65-L74 
What do you think?

Reply:
@rinx what do you think about this changes?
https://github.com/vdaas/vald/pull/862/commits/237248dac1be8035ea4095043226258029f18f6f
cc: @kpango

Reply:
good :+1:

Reply:
thank you :)